doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,206 @@
+                 |cGRE| PG +-+           +----+--------+
+                 +----+----+ |           |cBLU| active +-+
+                             |           +----+--------+ |
+                             V                           |
+                           +----+--------+               V
+                           |cBLU| osdmap +-+           +----+----------+
+                           +----+--------+ |           |cBLU| recovery +-+
+                                           |           +----+----------+ |
+                                           V                             |
+                                         +----+--------+                 V
+                                         |cBLU| active +-+             +----+---------+
+                                         +----+--------+ |             |cBLU| obj ctx +-+
+                                                         |             +----+---------+ |
+                                                         V                              |
+                                                       +----+----------+                V
+                                                       |cBLU| recovery +-+             +----+---------+
+                                                       +----+----------+ |             |cBLU| process |
+                                                                         |             +----+---------+
+                                                                         V
+                                                                       +----+---------+
+                                                                       |cBLU| obj ctx +-+
+                                                                       +----+---------+ |
+                                                                                        |
+                                                                                        V
+                                                                                       +----+---------+
+                                                                                       |cBLU| process |
+                                                                                       +----+---------+
+
+In the diagram above, two requests from the same client are being served in parallel. They share the same
+pipeline. But before one transits to the next stage, it have to wait until another leaves that stage.
+
+Better Throughput
+=================
+
+Some of us might argue, can we improve the throughput using some well known techniques?
+
+Reordering
+----------
+
+When serving requests from a client performing asynchronous I/O, OSD has the opportunity to execute
+multiple requests from the same client in parallel. But due to the second restriction, we should always
+reply the client in-order. But this does not prevent us from **executing** the requests ouf-of-order, as
```

Comment:
Nope, again, only a single object.  In general, a rados client will be sending requests on mutliple objects concurrently which mostly won't be even to the same OSD.

Reply:
Also, we are perfectly allowed to reorder a read ahead of a write on a particular object as long as the read isn't write ordered.

Reply:
We actually can't execute writes out of order from a particular client on a particular object (at least, not for any reasonably definition of execute).